### PR TITLE
Allow parallel execution of public unittests

### PR DESCRIPTION
**Pull request by wilzbach**
This allows to run the execution of the independent public unittests (aka ddoc-ed) in parallel with GNUMake (e.g. `-j16).

This is a first step on the journey of cleaning up the `Makefile`.

CC @jmdavis @schveiguy

### CODE REVIEW

File: posix.mak [hunk 1]
```diff
@@ -479,8 +488,11 @@ CWS_TOCHECK += $(ALL_D_FILES) index.d
 checkwhitespace: $(LIB) $(TOOLS_DIR)/checkwhitespace.d
 	$(DMD) $(DFLAGS) -defaultlib= -debuglib= $(LIB) -run $(TOOLS_DIR)/checkwhitespace.d $(CWS_TOCHECK)
 
-$(TOOLS_DIR)/checkwhitespace.d:
-	git clone --depth=1 ${GIT_HOME}/tools $(TOOLS_DIR)
+${TOOLS_DIR}:
+	git clone --depth=1 ${GIT_HOME}/$(@F) $@
+$(TOOLS_DIR)/checkwhitespace.d: | $(TOOLS_DIR)
+$(TOOLS_DIR)/styles/tests_extractor.d: | $(TOOLS_DIR)
+$(TOOLS_DIR)/styles/has_public_example.d: | $(TOOLS_DIR)
```
Comment (CyberShadow) [inline posix.mak]: Would have been nice to put this in a separate commit

File: posix.mak [hunk 2]
```diff
@@ -251,6 +251,15 @@ MAKEFILE = $(firstword $(MAKEFILE_LIST))
 # build with shared library support (defaults to true on supported platforms)
 SHARED=$(if $(findstring $(OS),linux freebsd),1,)
 
+# Check independent executability of public unittests on dlang.org
+IGNORED_PUBLICTESTS= $(addprefix std/, \
```
Comment (CyberShadow) [inline posix.mak]: I think this needs a more detailed comment of what this does and why it's needed

File: posix.mak [hunk 3]
```diff
@@ -547,12 +559,14 @@ style: ../dscanner/dsc $(LIB) has_public_example publictests
...
-	# execute all parsed tests
-	for file in $$(find out -name '*.d'); do echo "executing $${file}" && $(DMD) $(DFLAGS) -defaultlib= -debuglib= $(LIB) -main -unittest -run $$file || exit 1 ; done
+publictests: $(PUBLICTESTS)
+
+$(TEST_EXTRACTOR): $(TOOLS_DIR)/styles/tests_extractor.d
+	DFLAGS="$(DFLAGS) $(LIB) -defaultlib= -debuglib= $(LINKDL)" $(DUB) build --compiler=$${PWD}/$(DMD) --root=$(TOOLS_DIR)/styles -c tests_extractor
+
+%.publictests: %.d $(LIB) $(TEST_EXTRACTOR)
+	@$(TEST_EXTRACTOR) --inputdir  $< --outputdir $(ROOT)
+	@$(DMD) $(DFLAGS) -defaultlib= -debuglib= $(LIB) -main -unittest -run $(ROOT)/$(subst /,_,$<)
```
Comment (CyberShadow) [inline posix.mak]: Not a regression but we could use more documentation for what this is and what it's used for here as well

File: posix.mak (hunk 1 above)
Comment (wilzbach) [inline posix.mak]: Ok. Thanks for the advice. Done
